Sets: Replace is_EmptySet by a new is_empty attribute

gschintgen opened the pull request:
Rationale: Checking for is_EmptySet can be achieved via isinstance(),
but it can't properly handle the unknown case, generally resulting
in "None" in the assumptions system. Also, a set may well be empty
even though it's not recognized as such by SymPy.

is_empty correctly returns True, False or None.

Specific code, particularly for the "False" case, is added to
EmptySet, Naturals, Naturals0, Rationals, Reals, Interval, Union.

#### References to other Issues or PRs
Fixes #16946.
Fixes #17471.

#### Other comments

No test problems are expected, since `is_EmptySet` wasn't actually used anywhere.

At various places throughout the codebase, there are checks for `is S.EmptySet`. As of now I did not change these instances even though the only case where `True` is returned by `is_empty` is for `S.EmptySet` i.e. if SymPy is already capable of recognizing an empty set.

More interesting is the case of `is not S.EmptySet`, those instances could potentially be logic errors: `not S.EmptySet` does not imply that the set is indeed non-empty.
```
# git grep "not S.EmptySet"
calculus/util.py:            if critical_values is not S.EmptySet:
functions/elementary/trigonometric.py:                    is not S.EmptySet and \
functions/elementary/trigonometric.py:                        7*S.Pi/2)) is not S.EmptySet:
functions/elementary/trigonometric.py:                    is not S.EmptySet:
functions/elementary/trigonometric.py:                        is not S.EmptySet:
sets/tests/test_fancysets.py:    assert q is not S.EmptySet
solvers/inequalities.py:                    if interval is not S.EmptySet:
solvers/inequalities.py:                if global_interval is not S.EmptySet:
solvers/solveset.py:        elif rhs_s is not S.EmptySet:
solvers/solveset.py:                            if new_value is not S.EmptySet:
solvers/solveset.py:                            if new_value is not S.EmptySet:
solvers/solveset.py:                    if soln is not S.EmptySet:
```

In any case, those would be issues unrelated to this PR. Comments?

#### Release Notes


<!-- BEGIN RELEASE NOTES -->
* sets
  * Added `is_empty` which returns `True`, `False` or `None` and deprecated `is_EmptySet`.
<!-- END RELEASE NOTES -->

gschintgen committed:
Sets: Replace is_EmptySet by a new is_empty attribute

Rationale: Checking for is_EmptySet can be achieved via isinstance(),
but it can't properly handle the unknown case, generally resulting
in "None" in the assumptions system. Also, a set may well be empty
even though it's not recognized as such by SymPy.

is_empty correctly returns True, False or None.

Specific code, particularly for the "False" case, is added to
EmptySet, Naturals, Naturals0, Rationals, Reals, Interval, Union.

oscargus commented:
What about keeping the name and add the better evaluation?

I think that if not, `is_EmptySet` should be kept and deprecated as there may be users that rely on it. (Not sure when it was introduced though...)

I wouldn't consider changing the behaviour from two to three possible return values as breaking, more like being in line with the rest of the code. So the underlying reason is clearly good! The code also looks good!

oscargus commented:
As I realize that `is_EmptySet` was never `False` it should be even less problematic to just correct the behaviour (the possibly challenging case otherwise would be dealing with `anone` for existing code but that is a non-issue). Still, `is_empty` appears to be a better name in some sense...

jksuom commented on sympy/sets/sets.py:
```diff
@@ -55,11 +55,12 @@ class Set(Basic):
     is_Complement = None
     is_ComplexRegion = False
 
+    is_empty = None
```
This could probably be `False`. Those places where it would have to be redefined are relatively few.

gschintgen replied:
In practice most sets that are not `EmpytSet` will indeed be False, but I think None is the safer choice. E.g. I didn't touch `Intersection` and it's still correct. As I see it there are quite some cases that could (theoretically) turn out to be unrecognized empty sets (Intersection, ProductSet, Complement). `None` is also safer if a new Set subclass is added later on and this property is forgotten / the default not overridden.

oscarbenjamin commented:
> What about keeping the name and add the better evaluation?

There is a convention in SymPy that `x.is_Capitalised` means `isinstance(x, Capitalized)` and is always True/False whereas `is_lowercase` queries some more generic property of an object and can be None if the property is unknown. There are a number of attributes throughout SymPy that don't meet this convention but I think it's something that we should clean up more generally.

On that basis `is_EmptySet` is fine. It should be defined as `False` in `Set` and overridden as `True` in `EmptySet`. However any code that uses `is_EmptySet` should probably be using a new `is_emptyset` attribute and should be adjusted to take into account the resulting ternary logic.

gschintgen committed:
Fix #17471: ImageSet with baseset=EmptySet; add ProductSet.is_empty

ImageSet with a constant/variable-free Lambda will check if baseset
is_empty before returning either EmptySet or FiniteSet.

ProductSet.is_empty is also added in this commit (otherwise one of
the existing tests fails).

gschintgen committed:
Fix compact case in Interval.is_empty; add UniversalSet.is_empty

- Compact case in Interval.is_empty can now actually be reached...
- Add proper test for it.
- Add UniversalSet.is_empty == False

gschintgen committed:
Reinstate .is_EmptySet as a deprecated Set property

gschintgen commented:
I've added a few commits (including commit messages).

@oscargus, @jksuom, @oscarbenjamin: Thanks for your comments! Here are my thoughts:

The current is_EmptySet is highly misleading, since e.g. `FiniteSet(1).is_EmptySet` returns `None` even though it's not of class `EmptySet`. Comparing this to `is_Integer` / `is_integer`, it's incoherent.
I've also now found out why it's actually unused in the codebase: It has been deliberately been taken out by @smichr in https://github.com/sympy/sympy/commit/6626e209c7a0a4958b4b5ef3f4e5b9b03ce40c66.

To protect SymPy users against unsuspected/sneaky logic changes I did not fix `is_EmptySet` (i.e. make it equivalent to `is S.EmptySet`) but reinstated it as a deprecated property. Is that Ok?

Besides currently being wrong, `is_EmptySet` is also unnecessary IMO: Whereas `x.is_Integer` is more readable and shorter than `isinstance(x, Integer)`, the same reasoning does not hold for singleton classes. `s is S.EmptySet` is perfectly readable and short enough. So why bother?

I also added a fix for the issue that made me look into EmptySets in the first place... (It actually checks for `is_empty == False`)

Also of note / for future reference:
current `if s is S.EmptySet` could be replaced by `if s.is_empty` without logic change;
current `if s is not S.EmptySet` could be replaced by `if not s.is_empty` without logic change;
to actually test for non-emptyness one can use `s.is_empty == False` (or `fuzzy_not(s.is_empty)`)

Since a direct `is` check is probably faster anyway, I didn't do a search&replace.

oscarbenjamin commented on sympy/sets/fancysets.py:
```diff
@@ -308,7 +312,10 @@ def __new__(cls, flambda, *sets):
+            if all(s.is_empty for s in sets):
+                return S.EmptySet
+            elif any(s.is_empty == False for s in sets):
+                return FiniteSet(flambda.expr)
```
ImageSet is very poorly defined - I plan to fix it. I'm not sure about the logic above though.

The situation here is that we have an ImageSet with a constant map. The result is empty set if the baseset is empty and FiniteSet otherwise. However I think we have to interpret the base set as the productset of the plural sets in which case if any is empty then the whole thing is empty:
```julia
In [11]: ProductSet(S.EmptySet, Interval(0, 1))                                                                                   
Out[11]: ∅
```
I think the best thing would be to use `self.baseset.is_empty` i.e. defer the logic to ProductSet. You can't really use `self.baseset` since this is a classmethod but it should be equivalent.

gschintgen replied:
Ouch, I was fixated on single-variable ImageSets and must have thought "Oh, it maps multiple sets to their image, ok then at least one must be non-empty". I completely forgot about multi-variable ImageSets and Lambdas...

I'll change it to:
```
        if not set(flambda.variables) & flambda.expr.free_symbols:
            if any(s.is_empty for s in sets):
                return S.EmptySet
            elif all(s.is_empty == False for s in sets):
                return FiniteSet(flambda.expr)
```

oscarbenjamin commented on sympy/sets/sets.py:
```diff
@@ -746,6 +753,11 @@ def __iter__(self):
+    @property
+    def is_empty(self):
+        if all(s.is_empty == False for s in self.sets):
+            return False
```
This should be any rather than all. Also maybe it makes sense to use `fuzzy_or`.

gschintgen replied:
Your comment had me scratching my head until I realized that I was constantly thinking about non-emptyness, while your comment is about emptyness. Since the property is `is_empty` it's indeed cleaner and simpler to `return fuzzy_or(s.is_empty ...)`.
I'll change it, thanks.

oscarbenjamin commented on sympy/sets/sets.py:
```diff
@@ -932,6 +944,19 @@ def right_open(self):
+    @property
+    def is_empty(self):
+        if self.start == S.NegativeInfinity or self.end == S.Infinity:
+            return False
```
Is the above check necessary? I don't like it because it implicitly assumes that we couldn't have an unevaluated closed interval like `Interval(-oo, -oo, left_open=False, right_open=False, evaluate=False)` (That's not currently possible but it *should* be)

gschintgen replied:
It has an impact on `Interval(x, oo).is_empty` (for general `x`).

Since currently Intervals are reals-only and SymPy should implicitly assume necessary conditions to hold (https://github.com/sympy/sympy/issues/17488#issuecomment-524037140), I made it evaluate unconditionally to `False` even if `x` hasn't specifically been declared real.

Without the mentioned checks `.is_empty` would return `None`.

Maybe that's for the best, i.e. remove the check and the corresponding failing test. In some contexts it may also be surprising to have `Interval(x, oo).is_empty` automatically evaluate to `False` even if `x == oo` is still a possibility.
I'll change it.

(Personally I'd also find it more elegant if the `is_extended_*` assumptions had a set-theoretic equivalent, so I'm in favor of supporting intervals of extended reals. But I won't claim to understand all of the implications it could have on other parts of the codebase.)

oscarbenjamin replied:
How about this:
```
def is_empty(self):
    if self.left_open or self.right_open:
        cond = self.start < self.end  # One/both bounds open
    else:
        cond = self.start <= self.end  # Both bounds closed
    return fuzzy_bool(cond)
```

oscarbenjamin replied:
Obviously I got the inequalities the wrong way round :)

gschintgen replied:
Somehow my answer by e-mail doesn't show up here. Anyway thanks, marking as resolved.

oscarbenjamin commented on sympy/sets/sets.py:
```diff
@@ -1102,6 +1127,11 @@ def _sup(self):
+    @property
+    def is_empty(self):
+        if any(set.is_empty == False for set in self.args):
+            return False
```
This is Union.is_empty so maybe it could be fuzzy_and

gschintgen replied:
I'll change it. Thanks.

gschintgen commented:
There's a test failure related to pickling in combination with the deprecation:
https://travis-ci.org/sympy/sympy/jobs/576025978#L592
I'm not sure about how to fix it.
(Add `ignore_warnings(SymPyDeprecationWarning)`?)

gschintgen committed:
Sets: .is_empty fixes and improvements (from code review)

gschintgen committed:
Sets: Add proper docstring for .is_empty property

Also fixes one typo and two broken LaTeX formulas.

oscarbenjamin commented on sympy/sets/fancysets.py:
```diff
@@ -308,7 +312,10 @@ def __new__(cls, flambda, *sets):
+            if any(s.is_empty for s in sets):
+                return S.EmptySet
+            elif all(s.is_empty == False for s in sets):
+                return FiniteSet(flambda.expr)
```
Can the two loops above be replaced by one using fuzzy_or?

gschintgen replied:
Yes, thanks! I still have to get used to some constructs.

oscarbenjamin commented on sympy/sets/sets.py:
```diff
@@ -55,11 +56,31 @@ class Set(Basic):
+        return None
+
+    @property
+    @deprecated(useinstead="is_empty", issue=16946, deprecated_since_version="1.5")
```
The alternative useinstead could be `is S.EmptySet` as you suggested

gschintgen replied:
I'll put in both. (`useinstead="is S.EmptySet or is_empty"`)

oscarbenjamin replied:
That's good. If the deprecation refers to #16946 we should make sure that the issue has a discussion of why you might want one or the other.

gschintgen replied:
I updated the discussion in #16946. Feel free to edit.

oscarbenjamin commented on sympy/sets/sets.py:
```diff
@@ -1466,9 +1503,14 @@ class EmptySet(with_metaclass(Singleton, Set)):
 
+    @property
+    @deprecated(useinstead="is_empty", issue=16946, deprecated_since_version="1.5")
+    def is_EmptySet(self):
```
Is this not inherited from `Set`?

gschintgen replied:
Unfortunately not, neither for `is_EmptySet = True` nor for a `def is_EmptySet(self)` property.

oscarbenjamin replied:
Why not? I just tried this with a new class and it seems to work:
```julia
In [4]: class A(Set): 
   ...:     pass 
   ...:                                                                                                                           

In [5]: A().is_EmptySet                                                                                                           

In [6]: A().is_EmptySet == None                                                                                                   
Out[6]: True
```

gschintgen replied:
I'm not sure I'm following: The default for the deprecated `is_EmptySet` of the parent class is `None`, so it must be overridden to `True` using either `is_EmptySet` or `def is_EmptySet()`. But as far as I can tell the deprecation is not inherited so I have to repeat it, hence the `@deprecated def is_EmptySet()`.

oscarbenjamin replied:
Oh sorry I see. You have to override it here to change the return value from None to True and then also need to add the deprecation. That makes sense.

oscarbenjamin commented on sympy/sets/tests/test_fancysets.py:
```diff
@@ -77,8 +77,10 @@ def test_integers():
+    assert ImageSet(Lambda(x, y), S.Integers) == {y}
+    assert ImageSet(Lambda(x, 1), S.EmptySet) == S.EmptySet
+    empty = Intersection(FiniteSet(log(2)/pi), S.Integers)
+    assert ImageSet(Lambda(x, 1), empty) != FiniteSet(1)  # issue #17471
```
What is the result here? Does it not evaluate? If so I think it would be better to use unchanged i.e.:
```python
assert unchanged(ImageSet, Lambda(x, 1), empty)
```

gschintgen replied:
Ok! I didn't want to let the test fail if `empty` is properly recognized as being empty in the future, but it's probably better to have the test be more specific.

oscarbenjamin commented on sympy/sets/tests/test_sets.py:
```diff
@@ -366,6 +409,13 @@ def test_ProductSet_of_single_arg_is_arg():
+    assert ProductSet(S.Integers, S.Reals).is_empty == False
+    empty = Intersection(FiniteSet(sin(2)/pi), S.Integers)
+    assert empty is not S.EmptySet and \
+            ProductSet(empty, S.Reals).is_empty == None
```
Rather than `is not S.EmptySet` should unchanged be used here?

gschintgen replied:
Well, I assume that I could put the following:
```
    assert unchanged(Intersection, FiniteSet(sin(2)/pi), S.Integers)
    empty = Intersection(FiniteSet(sin(2)/pi), S.Integers)
    assert ProductSet(empty, S.Reals).is_empty == None
```
but to me that seems to slightly obscure the intent of the test and it adds a separate assert that does not really fit in. Is there a better way to rewrite the test?

oscarbenjamin replied:
I think just adding the assert unchanged line and keeping the rest would be fine. Using `is not` or `!=` in asserts is generally a very weak test of correctness (unless you are specifically testing `__eq__/__ne__`).

Looking closer though I'm not sure about this test. Ideally this would evaluate to the empty set and is_empty would give True. Perhaps a better example would be `empty = Intersection(FiniteSet(x), FiniteSet(2))` since if x is an arbitrary symbol then that necessarily has `is_empty == None`.

gschintgen replied:
Hmm, theoretically that'd be fine too, but I'm just noticing this (with current master):
```
>>> Intersection(FiniteSet(x), FiniteSet(2))
{2}
```
That's quite an ugly bug.
Ha! You've even filed it already yourself :-)
https://github.com/sympy/sympy/issues/16953

So I think for now the test is good enough? I've added the "unchanged" assert though.

oscarbenjamin replied:
Okay so maybe that example isn't so good right now :)

How about  `Interval(x, 1)`?

gschintgen replied:
Ok, changed.

oscarbenjamin commented on sympy/utilities/tests/test_pickling.py:
```diff
@@ -166,8 +166,10 @@ def test_core_undefinedfunctions_fail():
+    # Set.is_EmptySet is deprecated since 1.5
+    with ignore_warnings(SymPyDeprecationWarning):
+        for c in (Interval, Interval(0, 2)):
+            check(c)
```
Not sure what this does but maybe the is_EmptySet should be fixed at source rather than ignoring the warnings here...

gschintgen replied:
What do you mean by "the is_EmptySet should be fixed at source"?

The `check` function ("Check that pickling and copying round-trips.") is rather technical (different pickling protocol versions etc.), so I didn't want to touch it and adding an ignore_warnings in there could potentially hide other bugs. As for is_EmptySet itself I could only remove the deprecation?

oscarbenjamin replied:
Presumably somewhere inside `check` the `is_EmptySet` attribute is being used. If this PR is deprecating `is_EmptySet` then we should make sure it's not used anywhere in SymPy.

oscarbenjamin replied:
I think that you can just add `is_EmptySet` to `excluded_attrs` at the top of the file.

oscarbenjamin replied:
The `excluded_attrs` list should have comments explaining the reasons e.g.:
```
excluded_attrs = set([
    '_assumptions',  # This is a local cache that isn't automatically filled on creation
    '_mhash',   # Cached after __hash__ is called but set to None after creation
    'message',   # This is an exception attribute that is present but deprecated in Py2 (can be removed when Py2 support is dropped
    'is_EmptySet',  # Deprecated from SymPy 1.5. This can be removed when is_EmptySet is removed.
])
```

gschintgen replied:
Thanks, I think those comments are definitely valuable.

jksuom commented on sympy/sets/sets.py:
```diff
@@ -60,7 +60,21 @@ class Set(Basic):
 
-    is_empty = None
+    @property
+    def is_empty(self):
```
The line above would probably do as well.

gschintgen replied:
I'm not sure I understand your comment. I replaced the variable `is_empty` by a property method in order to put in the docstring, so that `is_empty` becomes part of the official documentation.

jksuom replied:
I think that it can be a simple attribute here, not a property. Fetching an attribute takes less time than evaluating a property. `'is_empty'` is just a key in `__dict__`. Its implementation does not have to be the same in all subclasses of a class.

gschintgen replied:
But how do I add the docstring then? I'm under the impression that it's not possible for simple variables.

I really wouldn't want to replicate the situation as it is for `is_iterable`. There's code that checks for it (I encountered it somewhere in "calculus.py" and wanted to look up some details), but found that it wasn't part of the official SymPy `Set` (class) documentation. I had to grep through the page to find `is_iterable` documented in `ProductSet`!? I had to check the code just to find out if and how `is_iterable` is defined in general and for which set classes it's available. IMO that's not how documentation is supposed to work...

jksuom replied:
I think that it can be documented in the Attributes section of [class docstring](https://numpydoc.readthedocs.io/en/latest/format.html#class-docstring). The difference between a plain data attribute and a property attribute is not visible to the user.

oscarbenjamin replied:
We should make a place to document all of the `is_*` attributes/properties. They should probably have their own page in the docs.

gschintgen replied:
@jksuom Thanks for the suggestion. I git grepped and had a look at other occurrences of "Attributes" and how it looked in the Sphinx-docs. There are examples in the geometry module. https://docs.sympy.org/latest/modules/geometry/lines.html
Apparently the "Attributes" section is formatted as a table with mostly one short line per attribute, completely separate from other method attributes. That's rather disappointing. For me `is_empty` would fit right in with other properties, such as `is_open`, `is_disjoint`, etc.

gschintgen committed:
Sets: some more .is_empty improvements

gschintgen committed:
Sets: simpler test for ProductSet.is_empty == None

gschintgen committed:
Remove unneeded unchanged import in test_sets.py

oscarbenjamin commented:
If this is merged then a deprecation removal issue should be opened for `is_EmptySet`.

gschintgen committed:
pickling tests: Move deprecation handling for is_EmptySet.

Co-Authored-By: Oscar Benjamin <oscar.j.benjamin@gmail.com>

oscarbenjamin commented:
This looks good. +1 to merge if the tests pass

oscarbenjamin commented:
Good work! Merging

oscarbenjamin commented:
A new deprecation removal issue should be opened with the deprecation removal label so we remember to remove this in future

gschintgen commented:
Thanks @oscarbenjamin and @jksuom for reviewing!
